### DISCUSSION TITLE
A problem following instructions installing server 3.7.0

### DISCUSSION THREAD

**Discussion by gene96817** [jamulussoftware/jamulus#1322] [Documentation]
I am following the instructions at jamulus.io/wiki/Server-Linux    Running a "headless" server.

At step 3, it would be helpful to know (1) does it matter which directory the command should be run, and (2) what is the result of the command, or how to check that the command completed properly. It could be, the command will create a file ffff in directory dddd.

Running step 4, I get the error "Failed to enable unit: Unit file jamulus-headless.service does not exist."
From which directory should this command be issued? I haven't been able to guess where jamulus-headless.service should be found.  I'm stuck.

Looking ahead at step 5, I have experience with some command line options for 3.6.2. However, it would be helpful for new server installers to have some command line examples.  The options table could also use an explicit mention of the defaults if an option is not selected.

Thanks.

**Comment by gilgongo** [comment]
Now that the recommended method of setting up a headless server is with the binary install, we will be moving the manual installation instructions into a separate file in the repo. So with luck some of your questions won't come up, since things like paths and whatnot are all pre-determined. 

However, in common with most projects, Jamulus also has a problem when presenting documents like the one you are reading: what level of assumed knowledge should we cater for?

For example "_how to check that the command completed properly_" is something that is common to any service being managed by systemd (you use systemctl or journalctl). Should we explain that? (I think we do in fact? See the section headed "Test that it starts OK").

Uncharitably, there are some who are of the opinion that if you don't know how to install a systemd service on Linux then you shouldn't be running a headless server (since there are security and other issues to take care of as well). I'm not of that opinion, but it's a tough one.

**Reply by gilgongo** [reply]
But to answer your questions meanwhile:

> does it matter which directory the command should be run

No.

> how to check that the command completed properly

Assuming you are using `sudo systemctl start jamulus` to start the server as per the docs, then you can check if it's running OK with `sudo systemctl status jamulus`. Otherwise try `sudo journalctl | grep -i jamulus` 

>  It could be, the command will create a file ffff in directory dddd

It's doesn't unless you mean a lockfile.

> The options table could also use an explicit mention of the defaults if an option is not selected.

Good point, although I'm not sure how many actually generate defaults. As to examples I would hope the options table is fairly simple in that regard. Did you have any specific examples in mind?

> Running step 4, I get the error "Failed to enable unit: Unit file jamulus-headless.service does not exist.

This isn't expected. I can't reproduce it at my end. It seems to indicate the unit file wasn't created though. Might need to look at this as it may be a bug?

**Reply by gene96817** [reply]
@gilgongo   Thank you for your reply. I'll respond to each issue here.

> However, in common with most projects, Jamulus also has a problem when presenting documents like the one you are reading: what level of assumed knowledge should we cater for?

I propose Jamulus is for a computer literate musician. We shouldn't require the musician to be a software developer or a highly experienced Linux admin. 

> For example "how to check that the command completed properly" is something that is common to any service being managed by systemd (you use systemctl or journalctl). Should we explain that? 

To be explicit, we should say, after this step, the file jamulus-headless.service will appear in ddd directory. We need to specify the files that are created? This way, the user can confirm nothing is missing and there are no surprises. In my example, we don't have a clue that step 3 completed correctly and we don't know why step 4 will not run. 

> Uncharitably, there are some who are of the opinion that if you don't know how to install a systemd service on Linux then you shouldn't be running a headless server (since there are security and other issues to take care of as well). I'm not of that opinion, but it's a tough one.

It is the culture of some developers to create solutions or themselves or only people expert like them. Jamulus is a wonderful tool for musicians. Most musicians are not developers. One choice is to say, Jamulus is not an end user product and musicians who are not developers should use a commercial Jamulus service. I hope this is not a serious choice. Let's make get even novice users to be successful installing a server. Security and other issues is a lesson for another day. The documentation could say, security and other issues should be part of a complete installation. It would be then the assignment for the user to research.

> Assuming you are using sudo systemctl start jamulus to start the server as per the docs, then you can check if it's running OK with sudo systemctl status jamulus. Otherwise try sudo journalctl | grep -i jamulus

From my experience with server 3.6.3, this will fail because jamulus-headless.service is missing. But that is a novice's guess.

> It could be, the command will create a file ffff in directory dddd
It's doesn't unless you mean a lockfile.

I was being too vague. In this example, step 3 should have created jamulus-headless.service in some directory. Where? What other files were touched or created? We should tell the user how to confirm the command (e.g. step 3) was properly done.

> This isn't expected. I can't reproduce it at my end. It seems to indicate the unit file wasn't created though. Might need to look at this as it may be a bug?

A bug? Oh no! Does this mean I can't install 3.7.0? Because of the changes in the repository, my 3.6.2 install scripts are also broken. I'll have to figure out how to update these scripts (written by a friend). Sadly I don't have a method to revert back to a working image.

**Reply by ann0see** [reply]
Does /lib/systemd/system/jamulus-headless.service exist and can you run `jamulus-headless -h`? If that's the case you might need to run the enable command again. 

Also: have a look at the syslog (`journalctl`). Maybe it shows you what the problem is

**Comment by gilgongo** [comment]
> I propose Jamulus is for a computer literate musician. We shouldn't require the musician to be a software developer or a highly experienced Linux admin.

I agree. But please consider that you don't also don't need to run a server (let alone a headless one on Linux) to use Jamulus. That's entirely optional.

> we should say, after this step, the file jamulus-headless.service will appear in ddd directory.

OK, but this would assume that if the user doesn't know how systemd works, then we need to explain all the attendant concepts to that as well, since systemd expects /lib/systemd/system/ is where all such unit files live by convention.

> we don't have a clue that step 3 completed correctly

In Unix, commands tend to complete silently if they are successful, again by convention. We could put a message in the installer, but no other installers do that to my knowledge. BTW if you are on Debian or Ubuntu, you can run `sudo updatedb` and `locate jamulus` to find all files with that name on the system.

> could say, security and other issues should be part of a complete installation

Beyond a general statement, specifics would be entirely dependent on what security environment they have, and there are too many variations of this.

**Reply by gene96817** [reply]
Thank you for your replies.

@ann0see    
/lib/systemd/system/jamulus-headless.service    does not exist

@gilgongo 

> But please consider that you don't also don't need to run a server (let alone a headless one on Linux) to use Jamulus. That's entirely optional.

Your point is taken. I installed the first server within 5600 km. The next nearest server is just too far. Also my chorus is too big to simply borrow an empty server.

>then we need to explain all the attendant concepts to that as well, since systemd expects /lib/systemd/system/ is where all such unit files live by convention.

When I get 3.7.0 running, I'll propose a solution about systemd. For now, /lib/systemd/system/  shows no signs of jamulus anything.

> you can run sudo updatedb and locate jamulus to find all files with that name on the system.

Thank you. I have Ubuntu installed. I find:
/usr/bin/jamulus
/usr/share/applications/jamulus.desktop
/usr/sharee/doc/jamulus
/usr/share/doc./jamulus/changelog.Debian.gz
/usr/share/doc/jamulus/copyright
/usr/share/doc/icons/hicolor/512x512/apps/jamulus.png
/var/lib/dpkg/info/jamulus.list
/var/lib/dpkg/info/jamulus.md5sums

Certainly nothing looking like jamulus-headless.service
I take the point that commands tend to complete silently. That is a good thing. However, in this case, I am wishing to know what failed.

What should be my next step?

**Reply by gilgongo** [reply]
OK when I try the instructions I see they tell you to download jamulus_headless_3.7.0 but then says to install it with `sudo apt install /path/to/jamulus_3.7.0`

Maybe that's the problem if you had non-headless installer in the same dir then tried to run `systemctl enable jamulus-headless` - the installer wouldn't have created the unit file since I assume only the server installer does that?

Either way, step 3 should say:

Install the package: sudo apt install /path/to/jamulus_**headless**_3.7.0_ubuntu_amd64.deb

**Reply by gene96817** [reply]
ooo.... I may have also interpreted step 3 improperly.
Does the "sudo apt install"  say that the file is in /path/to/jamulus_headless_3.7.0_ubuntu_amd64.deb?
I interpreted it to say the file is in `<path to>`/jamulus_headless_3.7.0_ubuntu_amd64.deb.
If it is the former, then I have to move jamulus_headless_3.7.0_ubuntu_amd64.deb to /path/to.

Please clarify.   Thanks

**Reply by ann0see** [reply]
@gilgongo propably I used the wrong Jekyll variable there. Can you change it on release for all languages?

**Reply by gene96817** [reply]
Progress report:

1 Download 
      The file should be (must be) saved in /path/to.
      I suggest this be stated explicitly for step 1.
      
3- Install the package
      Yes I should have believed the instructions were correct and an experienced reader would have known the file should have been saved into /path/to. Now there is also lots of output regarding the installation.

4- Enable the headless server.....
     Now the enable happens with some output.
    Now there is a file /lib/systemd/system/jamulus-headless.service

Thanks for clues.
I am now off to step 5.

**Reply by ann0see** [reply]
Ok. Maybe once you set it up, you can open a Pull Request to the changes branch for changing the process to a more user friendly way

**Reply by gilgongo** [reply]
> The file should be (must be) saved in /path/to.

At the risk of me appearing silly, you don't _literally_ mean a path called "/path/to" do you? That's just for purposes of the docs, we're not expecting you to actually create it. Some install guides say you should cd to /tmp though, which may be clearer?

**Reply by gene96817** [reply]
Well, I originally thought it couldn't have been literally /path/to. BUT after your comments, I moved to path/to/jamulus_headless_3.7.0_ubuntu_amd64.deb AND step 3 and step 4 worked. So both you and I were wrong to think it was silly. Maybe we want to change the location? /etc/jamulus/jamulus_headless_3.7.0_ubuntu_amd64.deb?

**Comment by gene96817** [comment]
Is there something illformed with this ExecStart from my jamulus-headless.service?

> ExecStart=/bin/sh -c '/usr/bin/jamulus-headless -e classical.jamulus.io:22524 -F -o HI-Studio1;Honolulu;225 -s -T -u 100 -w /etc/jamulus/jamulus_welcome -n --serverpublicip 72.253.194.252'

Not all the parameters are seen in systemctl status jamulus-headless. Does the status look right? The server is not registering with the directory (or the directory server is not showing this server).

> root@gene-MS-7C77:/home/gene# sudo systemctl status jamulus-headless
● jamulus-headless.service - Jamulus headless server
     Loaded: loaded (/lib/systemd/system/jamulus-headless.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2021-03-23 15:57:46 HST; 15min ago
   Main PID: 2838 (sh)
      Tasks: 4 (limit: 38288)
     Memory: 1.7M
     CGroup: /system.slice/jamulus-headless.service
             ├─2838 /bin/sh -c /usr/bin/jamulus-headless -e classical.jamulus.io:22524 -F -o HI-Studio1;Honolulu;>
             └─2839 /usr/bin/jamulus-headless -e classical.jamulus.io:22524 -F -o HI-Studio1

> Mar 23 15:57:46 gene-MS-7C77 systemd[1]: Started Jamulus headless server.
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]: - central server: classical.jamulus.io:22524
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]: - using 64 samples frame size mode
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]: - server info: HI-Studio1
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]: No GUI support compiled. Running in headless mode.
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]:  *** Jamulus, Version 3.7.0
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]:  *** Internet Jam Session Software
Mar 23 15:57:46 gene-MS-7C77 jamulus[2839]:  *** Released under the GNU General Public License (GPL)
lines 1-18/18 (END)
